cclib/method/moments.py

```diff
@@ -0,0 +1,131 @@
+        dipole = numpy.dot(charges, transl_coords_au)
+        
+        return convertor(dipole, 'ebohr', 'Debye')
+
+    def _calculate_quadrupole(self, charges, coords, origin):
+        """Calculate the traceless quadrupole moment from the given
+        atomic charges and their coordinates with respect to the origin.
+        """
+        transl_coords_au = convertor(coords - origin, 'Angstrom', 'bohr')
+
+        delta = numpy.eye(3)
+        Q = numpy.zeros([3, 3])
+        for i in range(3):
+            for j in range(3):
+                for q, r in zip(charges, transl_coords_au):
+                    Q[i,j] += 1/2 * q * (3 * r[i] * r[j] - \
+                              numpy.linalg.norm(r)**2 * delta[i,j])
+
+        triu_idxs = numpy.triu_indices_from(Q)
+        raveled_idxs = numpy.ravel_multi_index(triu_idxs, Q.shape)
+        quadrupole = numpy.take(Q.flatten(), raveled_idxs)
+
+        return convertor(quadrupole, 'ebohr2', 'Buckingham')
+    
+    def calculate(self, origin='nuccharge', population='mulliken',
+                  masses=None):
+        """Calculate electric dipole and quadrupole moments using parsed
+        partial atomic charges.
+        
+        Inputs:
+            origin - a choice of the origin of coordinate system. Can be
+                either a three-element iterable or a string. If
+                iterable, then it explicitly defines the origin (in
+                Angstrom). If string, then the value can be any one of
+                the following and it describes what is used as the
+                origin:
+                    * 'nuccharge' -- center of positive nuclear charge
+                    * 'mass' -- center of mass
+            population - a type of population analysis used to extract
+                corresponding atomic charges from the output file.
+            masses - if None, then use default atomic masses. Otherwise,
+                the user-provided will be used.
+
+        Returns:
+            A list where the first element is the origin of coordinates,
+            while other elements are dipole and quadrupole moments
+            expressed in terms of Debye and Buckingham units
+            respectively.
+
+        Notes:
+            To calculate the quadrupole moment the Buckingham definition
+            [1]_ is chosen. Hirschfelder et al. [2]_ define it two times
+            as much.
+
+        References:
+         .. [1] Buckingham, A. D. (1959). Molecular quadrupole moments.
+            Quarterly Reviews, Chemical Society, 13(3), 183.
+            https://doi.org:10.1039/qr9591300183.
+         .. [2] Hirschfelder J. O., Curtiss C. F. and Bird R. B. (1954).
+            The Molecular Theory of Gases and Liquids. New York: Wiley.
+        """
+        coords = self.data.atomcoords[-1]
+        try:
+            charges = self.data.atomcharges[population]
+        except KeyError:
+            raise ValueError
```

Comment:
Why is it useful to catch this error and recast it as a different type? If we want to go in this direction, please make a custom exception type and use that here for clarity.

Also, consider doing `except KeyError as e` and pass `e` along to the new exception.

Also, please document this in a Raises section of the docstring.

Reply:
> Why is it useful to catch this error and recast it as a different type? 

Since `KeyError`and `AttributeError` don't semantically describe the true cause of the error connected with the inappropriate inputs to our method.